### PR TITLE
fix: kubectl expose fails for apps with same-port, different-protocol

### LINKED ISSUE #114402: `kubectl expose` fails for apps with same-port, different-protocol
Given this deployment:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: multiport
spec:
  replicas: 0
  selector:
    matchLabels:
      app: multiport
  template:
    metadata:
      labels:
        app: multiport
    spec:
      containers:
      - image: k8s.gcr.io/serve_hostname:v1.4
        name: c1
        ports:
        - containerPort: 93
          protocol: TCP
        - containerPort: 93
          protocol: UDP
```

```
$ k expose deploy multiport
The Service "multiport" is invalid: spec.ports[1]: Duplicate value: core.ServicePort{Name:"", Protocol:"UDP", AppProtocol:(*string)(nil), Port:93, TargetPort:intstr.IntOrString{Type:0, IntVal:0, StrVal:""}, NodePort:0}
```

Looking at staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go, it builds a `protocolsMap` which reduces those 2 ports to one protocol.  It should probably just return a list of {proto, port} or even {proto, port, name} which it extracts from the target.

### PR DESCRIPTION
**Pull request by aimuz**
Fixed: https://github.com/kubernetes/kubernetes/issues/114402
Signed-off-by: aimuz <mr.imuz@gmail.com>

#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #114402

#### Does this PR introduce a user-facing change?

```release-note
Kubectl expose supports the creation of different protocol service on the same port
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by aimuz** [comment]
/assign anggao

Can you review the code for me

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go [hunk 1]
```diff
@@ -390,15 +388,15 @@ func (o *ExposeServiceOptions) createService() (*corev1.Service, error) {
...
 	}
 
-	name := o.Name
-	if len(name) == 0 {
-		name = o.DefaultName
-		if len(name) == 0 {
+	serviceName := o.Name
+	if len(serviceName) == 0 {
+		serviceName = o.DefaultName
+		if len(serviceName) == 0 {
```
Comment (aimuz) [inline staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:394]: This is to avoid the misunderstanding of duplicate names. In the loop, name is also used. To avoid misunderstanding, I renamed it

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go [hunk 2]
```diff
@@ -442,10 +437,23 @@ func (o *ExposeServiceOptions) createService() (*corev1.Service, error) {
...
+						})
+					}
+					continue
 				}
 			}
+
+			name := servicePortName
+			if len(portStringSlice) > 1 {
+				name = fmt.Sprintf("%s-%d", strings.ToLower(protocol), port)
+			}
```
Comment (aimuz) [inline staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:456]: Compared with the naming of prot-i, this method may be easier to understand, which is also to distinguish between different protocols on the same port

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose_test.go [hunk 3]
```diff
@@ -1722,8 +1756,8 @@ func TestGenerateService(t *testing.T) {
 			service, err := exposeServiceOptions.createService()
 			if test.expectErr == "" {
 				require.NoError(t, err)
-				if !apiequality.Semantic.DeepEqual(service, test.expected) {
-					t.Errorf("\nexpected:\n%#v\ngot:\n%#v", test.expected, service)
+				if diff := cmp.Diff(test.expected, service); diff != "" {
```
Comment (aimuz) [inline staging/src/k8s.io/kubectl/pkg/cmd/expose/expose_test.go:1759]: This will output more intuitive differences

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go [hunk 4]
```diff
@@ -374,184 +370,26 @@ func (o *ExposeServiceOptions) RunExpose(cmd *cobra.Command, args []string) erro
...
+		PortName:        o.PortName,
+		TargetPort:      o.TargetPort,
+		ExternalIP:      o.ExternalIP,
+		Type:            o.Type,
+		LoadBalancerIP:  o.LoadBalancerIP,
+		SessionAffinity: o.SessionAffinity,
+		ClusterIP:       o.ClusterIP,
 	}
-	selector, err := parseLabels(o.Selector)
+	obj, err := versioned.GenerateService(opts)
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:390]: This looks unusual. What is the reason of this change?

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go [hunk 5]
```diff
@@ -374,184 +370,26 @@ func (o *ExposeServiceOptions) RunExpose(cmd *cobra.Command, args []string) erro
...
-		if len(portProtocol[0]) == 0 {
-			return nil, fmt.Errorf("unexpected empty port")
-		}
-		if len(portProtocol[1]) == 0 {
-			return nil, fmt.Errorf("unexpected empty protocol")
-		}
-		portProtocolMap[portProtocol[0]] = portProtocol[1]
-	}
-	return portProtocolMap, nil
+	return obj, nil
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:394]: What is the reason of these changes with respect to the referenced issue?

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/interface.go [hunk 6]
```diff
@@ -67,7 +67,7 @@ var MapBasedSelectorForObjectFn MapBasedSelectorForObjectFunc = mapBasedSelector
 
 // ProtocolsForObjectFunc will call the provided function on the protocols for the object,
 // return nil-map if no protocols for the object, or return an error.
-type ProtocolsForObjectFunc func(object runtime.Object) (map[string]string, error)
+type ProtocolsForObjectFunc func(object runtime.Object) (map[string][]string, error)
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/polymorphichelpers/interface.go:70]: Is this backward compatible?

**Review by ardaguclu — COMMENTED** [review]
Thanks for the fix and the effort you invested.

Maybe it is better to do less changes to fix that issue without touching generated files.

to prevent accidental merging;
/hold

**Comment by ardaguclu** [comment]
As far as I can see, integration test failure is real.

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go (hunk 4 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:390]: Repetitive code implementation, which I extract as a public method

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go (hunk 5 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:394]: Repetitive code implementation, which I extract as a public method

**Comment by aimuz** [comment]
> Maybe it is better to do less changes to fix that issue without touching generated files.

 It does not involve the modification of generated files. Their file names are special

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/interface.go (hunk 6 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/interface.go:70]: No, but I don't think this change will have any impact. Only kubectl will use this

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go (hunk 5 above)
Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:394]: Is this something required for fixing the bug or is this refactoring?.

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/interface.go (hunk 6 above)
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/interface.go:70]: It is public and it looks like people are using it (via github search, but possibly more elsewhere)...
https://github.com/search?q=ProtocolsForObjectFunc&type=code

Can we leave the existing function as-is and create a new one with the new behavior?

Perhaps mark the old one as deprecated, with an explanation that it doesn't support multiple protocols per port, and that code should use the new function instead?

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject.go [hunk 7]
```diff
@@ -60,19 +61,21 @@ func portsForObject(object runtime.Object) ([]string, error) {
 }
 
 func getPorts(spec corev1.PodSpec) []string {
-	result := []string{}
+	result := sets.New[string]()
```
Comment (brianpursley) [inline staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject.go:64]: Does this need to change as part of this PR, or can it be done separately?  I'm not saying it is wrong, but it is a subtle change in the output (if there are duplicate ports) and potentially the order of the output.  I'm not necessarily opposed to this change, just think that maybe it should be its own PR and would need some unit tests as well.

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go [hunk 8]
```diff
@@ -19,25 +19,23 @@ package expose
...
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructuredscheme"
 	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/cli-runtime/pkg/genericclioptions"
 	"k8s.io/cli-runtime/pkg/printers"
 	"k8s.io/cli-runtime/pkg/resource"
 	cmdutil "k8s.io/kubectl/pkg/cmd/util"
 	"k8s.io/kubectl/pkg/generate"
+	"k8s.io/kubectl/pkg/generate/versioned"
```
Comment (brianpursley) [inline staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:38]: We don't want to use generators, we want to reduce and/or eliminate their usage.

In fact, this same import was removed in 1.24:
https://github.com/kubernetes/kubernetes/pull/106824

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go (hunk 4 above)
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:390]: See my comment above.  We don't want to use the generators.

**Review by brianpursley — CHANGES_REQUESTED** [review]
Hi @aimuz, thanks for taking on this task.

As mentioned in some of the review comments, can you limit the PR to fixing the bug, without the refactoring for now?  I think it is a good place to start, then refactoring can be done later if needed (either as part of this PR, if it makes sense, or as a new PR, which often is the case).

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go (hunk 8 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:38]: Thank you, I will modify it

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject.go (hunk 7 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject.go:64]: I think this change is necessary because duplicate ports need to be eliminated. There should also be no external scheduling for this method. As for the unit test, I will add it to confirm whether there is a sequence problem.

**Comment by aimuz** [comment]
Thank you. Your suggestion is necessary. I will readjust it. I'm sorry I don't know the delete of the generate plan. Next, I will correct it

@brianpursley @ardaguclu

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go [hunk 9]
```diff
@@ -122,7 +120,7 @@ type ExposeServiceOptions struct {
 	CanBeExposed              polymorphichelpers.CanBeExposedFunc
 	MapBasedSelectorForObject func(runtime.Object) (string, error)
 	PortsForObject            polymorphichelpers.PortsForObjectFunc
```
[the remark this answers is not in the document]
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:123]: Yes, we still need

**Comment by aimuz** [comment]
I made an update. Can you have a look

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go [hunk 10]
```diff
@@ -442,8 +439,20 @@ func (o *ExposeServiceOptions) createService() (*corev1.Service, error) {
...
 				protocol = "TCP" // have the default so we can stay sane
-				if exposeProtocol, found := portProtocolMap[stillPortString]; found {
-					protocol = exposeProtocol
+				if exposeProtocols, found := portProtocolMap[stillPortString]; found {
+					if len(exposeProtocols) == 1 {
+						protocol = exposeProtocols[0]
+						break
+					}
+					for _, exposeProtocol := range exposeProtocols {
+						name := fmt.Sprintf("port-%d-%s", i+1, strings.ToLower(exposeProtocol))
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:448]: What is the reason of this name generation mechanism?

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject_test.go [hunk 11]
```diff
@@ -113,28 +194,31 @@ func TestPortsForObject(t *testing.T) {
...
 		},
 	}
-	expectedPorts := []string{"101"}
 
 	for _, test := range tests {
-		actual, err := portsForObject(test.object)
-		if test.expectErr {
-			if err == nil {
-				t.Error("unexpected non-error")
+		t.Run(test.name, func(t *testing.T) {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject_test.go:206]: I understand the purpose in here. But considering that there are several changes for the fix, it would be better not touch unit tests(unless it is essential) to better verify the actual changes.

File: staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go (hunk 10 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go:448]: The generation rule of this name is to be consistent with the existing rules

 The current rule is 

```
name := fmt.Sprintf("port-%d", i+1）
```

In order to solve the problem of duplication, the protocol will be added later


```
name := fmt.Sprintf("port-%d-%s", i+1, protocol)
```

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject_test.go (hunk 11 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject_test.go:206]: Thank you. Let me withdraw this change. Can I add a PR to complete the change here?

**Comment by aimuz** [comment]
@brianpursley @ardaguclu 
Can you help me review the code again to see if there are any problems with the code

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject_test.go [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: map[string][]string{"101": {"TCP"}},
+		},
+		{
+			name:      "unsupported object",
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject_test.go:188]: Does `portsforobject` also return `expectErr` true?

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject.go [hunk 13]
```diff
@@ -60,19 +60,31 @@ func portsForObject(object runtime.Object) ([]string, error) {
 }
 
 func getPorts(spec corev1.PodSpec) []string {
-	result := []string{}
+	var result []string
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/polymorphichelpers/portsforobject.go:63]: We are already introducing a new `multiprotocolsforobjects` functionality, why do we need to make changes on `portsforobject`?

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+	}
+}
+
+func getMultiProtocols(spec corev1.PodSpec) map[string][]string {
+	result := make(map[string][]string)
+	for _, container := range spec.Containers {
+		for _, port := range container.Ports {
+			// Empty protocol must be defaulted (TCP)
+			if len(port.Protocol) == 0 {
+				port.Protocol = corev1.ProtocolTCP
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject.go:70]: Although `port.Protocol` is value type(it looks safe), I think, we should avoid such assignments to objects inside `corev1.PodSpec`.
Instead, I'd rather prefer;
```go
protocol := corev1.ProtocolTCP
if len(port.Protocol) > 0 {
protocol = port.Protocol
}
```

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject.go [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+	return result
+}
+
+// Extracts the protocols exposed by a service from the given service spec.
+func getServiceMultiProtocols(spec corev1.ServiceSpec) map[string][]string {
+	result := make(map[string][]string)
+	for _, servicePort := range spec.Ports {
+		// Empty protocol must be defaulted (TCP)
+		if len(servicePort.Protocol) == 0 {
+			servicePort.Protocol = corev1.ProtocolTCP
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject.go:85]: same above

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject_test.go (hunk 12 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject_test.go:188]: yes

This is the code for portsforobject

```go
func portsForObject(object runtime.Object) ([]string, error) {
	switch t := object.(type) {
	case *corev1.ReplicationController:
		return getPorts(t.Spec.Template.Spec), nil

	case *corev1.Pod:
		return getPorts(t.Spec), nil

	case *corev1.Service:
		return getServicePorts(t.Spec), nil

	case *extensionsv1beta1.Deployment:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1.Deployment:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1beta2.Deployment:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1beta1.Deployment:
		return getPorts(t.Spec.Template.Spec), nil

	case *extensionsv1beta1.ReplicaSet:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1.ReplicaSet:
		return getPorts(t.Spec.Template.Spec), nil
	case *appsv1beta2.ReplicaSet:
		return getPorts(t.Spec.Template.Spec), nil
	default:
		return nil, fmt.Errorf("cannot extract ports from %T", object)
	}
}
```

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject.go (hunk 14 above)
Comment (aimuz) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/multiprotocolsforobject.go]: @ardaguclu done

**Comment by ardaguclu** [comment]
Thanks @aimuz 

/lgtm
/triage accepted
/priority backlog

I'm deferring to @brianpursley for last approval, if he wants to check.

**Comment by aimuz** [comment]
/pin @brianpursley

Do you have time for a code review, thank you

**Comment by aojea** [comment]
great 👏

**Comment by aimuz** [comment]
Thank you all for your hard work!

**Comment by aimuz** [comment]
Does this need to be Cherry-picked to another branch?

@ardaguclu @aojea
